### PR TITLE
[AP] Re-Designed the Legalizer Abstraction

### PR DESCRIPTION
**Pull request by AlexandreSinger**
Abstracted the tile locations into a TileGraph class.

Abstracted the architecture information into an Architecture Model class.

This creates an easier interface to extend on in the future, as we make the legalizer heterogenous and more powerful.

### CODE REVIEW

**Comment by AlexandreSinger** [comment]
@robluo Please review.

File: vpr/src/place/analytical_placement/PlacementLegalizer.cpp [hunk 1]
```diff
@@ -2,106 +2,541 @@
...
+            return overflow_sub_tile;
+        // TODO: Definately make these ones debug.
+        VTR_ASSERT(!sub_tiles.empty());
+        VTR_ASSERT(sub_tile_idx < sub_tiles.size());
+        // Else, return the subtile.
+        return sub_tiles[sub_tile_idx];
+    }
+
+    // Add an atom to a given sub_tile.
+    void add_atom_to_sub_tile(size_t node_id, size_t sub_tile_idx) {
```
Comment (robluo) [inline vpr/src/place/analytical_placement/PlacementLegalizer.cpp:98]: Should these two API support a way for caller to know if the atom exists already in the `add` function, or it does not exist in the `remove` function?

File: vpr/src/place/analytical_placement/PlacementLegalizer.cpp [hunk 2]
```diff
@@ -2,106 +2,541 @@
...
+                    device.grid.get_height_offset(tile_loc) != 0) {
+                    continue;
+                }
+                PlaceTile &place_tile = tiles[tile_grid[x][y]];
+                size_t tile_width = place_tile.physical_tile_type.width;
+                size_t tile_height = place_tile.physical_tile_type.height;
+                // Add the neighbors.
+                std::unordered_set<PlaceTileId> neighbor_tile_ids;
+                // Add unique tiles on left and right sides
+                for (size_t ty = y; ty < y + tile_height; ty++) {
```
Comment (robluo) [inline vpr/src/place/analytical_placement/PlacementLegalizer.cpp:181]: I'm probably wrong, but shouldn't there only be at most 4 neighbouring tiles?

**Comment by robluo** [comment]
LGTM

**Pull request merged**